File: backend/fill_db.py
from datetime import date
from sqlmodel import SQLModel

from app.domain.track import Track
from app.domain.user import User
from app.domain.exercise import Exercise, ExerciseSlug
from app.domain.workout import Workout, WorkoutSection, WorkoutProtocol, WorkoutExercise

from app.infrastructure.db.database import db
from app.infrastructure.db.user import UserDbRepository
from app.infrastructure.db.track import TrackDbRepository
from app.infrastructure.db.exercise import ExerciseDbRepository
from app.infrastructure.db.workout import WorkoutDbRepository

session = next(db.session_getter())

# Clear database
SQLModel.metadata.drop_all(session.get_bind())
SQLModel.metadata.create_all(session.get_bind())

session.begin()

# create user
user_repo = UserDbRepository(session)

user_login = "first_user"
user_id = User.create_id(user_login)
user = User(
    id=user_id, name="First User", login=user_login, email="first_user@example.com"
)

print(f"create user: {user}")
user_repo.add(user)

# create track
track_repo = TrackDbRepository(session)

track = Track(name="Нескучный спорт", owner_id=user_id)

print(f"create track: {track}")
track_repo.add(track)

# create exercises
exercise_repo = ExerciseDbRepository(session)
    
exercises = {
    'Раскрытия в планке': Exercise(name="Раскрытия в планке", slug=ExerciseSlug("plank-hip-opening"), description=""),
    'Ягодичные марши': Exercise(name="Ягодичные марши", slug=ExerciseSlug("glute-march"), description=""),
    'Джампинг джек': Exercise(name="Джампинг джек", slug=ExerciseSlug("jumping-jack"), description=""),
    'C пола на грудь + 2 выпада назад': Exercise(name="C пола на грудь + 2 выпада назад", slug=ExerciseSlug("push-up-with-back-drop"), description=""),
    'Становая на одной': Exercise(name="Становая на одной", slug=ExerciseSlug("deadlift-on-one-leg"), description=""),
    'Пресс на прямых руки над головой': Exercise(name="Пресс на прямых руки над головой", slug=ExerciseSlug("situps-with-hands-over-head"), description=""),
    
    'Стол': Exercise(name="Стол", slug=ExerciseSlug("table"), description=""),
    'Наклоны вперед': Exercise(name="Наклоны вперед", slug=ExerciseSlug("forward-bend"), description=""),
    'Качающиеся планки': Exercise(name="Качающиеся планки", slug=ExerciseSlug("plank-with-jumping-jack"), description=""),
    'Становая тяга': Exercise(name="Становая тяга", slug=ExerciseSlug("deadlift"), description=""),
    'Приседания': Exercise(name="Приседания", slug=ExerciseSlug("squats"), description=""),
    'С груди над головой': Exercise(name="С груди над головой", slug=ExerciseSlug("push-up-with-hands-over-head"), description=""),
}

for exercise in exercises.values():
    print(f"create exercise: {exercise}")
    exercise_repo.add(exercise)

# create workout
workout_repo = WorkoutDbRepository(session)

"""
31 Jan
Разминка
3 раунда:
- 20 раскрытий в планке
- 20 ягодичных маршей
- 30 джампинг Джеков

Комплекс
5 раундов не на время:
- 10 с пола на грудь + 2 выпада назад*
- 10+10 становая на одной*
- 10 пресс на прямых руки над головой*
"""

workout = Workout(
    date=date(2025, 1, 31),
    track_id=track.id,

    sections=[
        WorkoutSection(
            title="Разминка", 
            protocol=WorkoutProtocol(title="3 раунда"),
            exercises=[
                WorkoutExercise(exercise_slug=exercises["Раскрытия в планке"].slug, repetitions=20),
                WorkoutExercise(exercise_slug=exercises["Ягодичные марши"].slug, repetitions=20),
                WorkoutExercise(exercise_slug=exercises["Джампинг джек"].slug, repetitions=30),
            ],
        ),
        WorkoutSection(
            title="Комплекс", 
            protocol=WorkoutProtocol(title="5 раундов"),
            description="*можно использовать спортивные снаряды",
            exercises=[
                WorkoutExercise(exercise_slug=exercises["C пола на грудь + 2 выпада назад"].slug, repetitions=10),
                WorkoutExercise(exercise_slug=exercises["Становая на одной"].slug, repetitions=20, repetitions_text="10+10"),
                WorkoutExercise(exercise_slug=exercises["Пресс на прямых руки над головой"].slug, repetitions=10),
            ],
        ),
    ]
)

print(f"create workout: {workout}")
workout_repo.add(workout)


"""
3 feb 
Разминка
3 раунда:
- 10 столов
- 10 наклонов вперед
- 20 качающихся планок на локтях

Комплекс
5 раундов на время:
- 24 становых*
- 18 приседаний*
- 12 с груди над головой*
"""

workout = Workout(
    date=date(2025, 2, 3),
    track_id=track.id,
    sections=[
        WorkoutSection(
            title="Разминка", 
            protocol=WorkoutProtocol(title="3 раунда"),
            exercises=[
                WorkoutExercise(exercise_slug=exercises["Стол"].slug, repetitions=10),
                WorkoutExercise(exercise_slug=exercises["Наклоны вперед"].slug, repetitions=10),
                WorkoutExercise(exercise_slug=exercises["Качающиеся планки"].slug, repetitions=20),
            ],
        ),
        WorkoutSection(
            title="Комплекс", 
            protocol=WorkoutProtocol(title="5 раундов"),
            exercises=[
                WorkoutExercise(exercise_slug=exercises["Становая на одной"].slug, repetitions=24),
                WorkoutExercise(exercise_slug=exercises["Приседания"].slug, repetitions=18),
                WorkoutExercise(exercise_slug=exercises["С груди над головой"].slug, repetitions=12),
            ],
        ),
    ]
)

print(f"create workout: {workout}")
workout_repo.add(workout)


# commit all changes
session.commit()

File: backend/scripts/fill_db/fill_db.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"time"

	"github.com/rs/zerolog"

	"github.com/healthy-heroes/neskuchka/backend/app/domain"
	"github.com/healthy-heroes/neskuchka/backend/app/storage/database"
)

func main() {
	logger := zerolog.New(os.Stdout).Level(zerolog.InfoLevel)

	// Database file path
	dbPath := "./bin/app.db"

	// Make sure directory exists
	dbDir := filepath.Dir(dbPath)
	if err := os.MkdirAll(dbDir, 0755); err != nil {
		log.Fatalf("Failed to create database directory: %v", err)
	}

	// Delete database file if it exists
	if _, err := os.Stat(dbPath); err == nil {
		if err := os.Remove(dbPath); err != nil {
			log.Fatalf("Failed to delete existing database file: %v", err)
		}
		fmt.Println("Existing database file deleted")
	}

	// Initialize DB
	engine, err := database.NewSqliteEngine(dbPath, logger)
	if err != nil {
		logger.Fatal().Err(err).Msg("Failed to connect to database")
	}
	defer engine.Close()

	dataStorage := database.NewDataStorage(engine, logger)

	// Begin transaction
	ctx := context.Background()
	tx, err := engine.Beginx()
	if err != nil {
		logger.Fatal().Err(err).Msg("Failed to begin transaction")
	}
	defer func() {
		if err != nil {
			tx.Rollback()
			return
		}
		err = tx.Commit()
		if err != nil {
			logger.Fatal().Err(err).Msg("Failed to commit transaction")
		}
	}()

	// Create user
	admin := domain.User{
		ID:    domain.NewUserID(),
		Name:  "Admin",
		Email: "admin@example.com",
	}
	_, err = dataStorage.CreateUser(ctx, admin)
	if err != nil {
		logger.Fatal().Err(err).Msg("Failed to create user")
	}

	// Create track
	mainTrack := domain.Track{
		ID:          domain.NewTrackID(),
		Name:        "Нескучный спорт",
		Slug:        domain.TrackSlug("main"),
		Description: "Тренируйтесь с нами — где бы вы ни находились!\nИдеальная программа, чтобы поддерживать форму дома, без специального оборудования.",
		OwnerID:     admin.ID,
	}
	_, err = dataStorage.CreateTrack(ctx, mainTrack)
	if err != nil {
		logger.Fatal().Err(err).Msg("Failed to create track")
	}

	// Create first workout (Jan 31, 2025)
	_, err = dataStorage.CreateWorkout(ctx, domain.Workout{
		ID:      domain.NewWorkoutID(),
		Date:    time.Date(2025, 1, 31, 0, 0, 0, 0, time.UTC),
		TrackID: mainTrack.ID,
		Notes:   "First workout",
		Sections: []domain.WorkoutSection{
			{
				Title: "Разминка",
				Protocol: domain.Protocol{
					Type:        domain.ProtocolTypeCustom,
					Title:       "3 раунда",
					Description: "",
				},
				Exercises: []domain.WorkoutExercise{
					{
						ExerciseSlug: domain.ExerciseSlug("plank-hip-opening"),
						Description:  "20 раскрытий в планке",
					},
				},
			},
			{
				Title: "Комплекс",
				Protocol: domain.Protocol{
					Type:        domain.ProtocolTypeCustom,
					Title:       "5 раундов",
					Description: "*можно использовать спортивные снаряды",
				},
				Exercises: []domain.WorkoutExercise{
					{
						ExerciseSlug: domain.ExerciseSlug("push-up-with-back-drop"),
						Description:  "10 отжиманий",
					},
					{
						ExerciseSlug: domain.ExerciseSlug("deadlift-on-one-leg"),
						Description:  "20 становых на одной",
					},
					{
						ExerciseSlug: domain.ExerciseSlug("situps-with-hands-over-head"),
						Description:  "10 пресса на прямых руках над головой",
					},
				},
			},
		},
	})
	if err != nil {
		logger.Fatal().Err(err).Msg("Failed to create workout")
	}

	// Create second workout (Feb 3, 2025)
	_, err = dataStorage.CreateWorkout(ctx, domain.Workout{
		ID:      domain.NewWorkoutID(),
		Date:    time.Date(2025, 2, 3, 0, 0, 0, 0, time.UTC),
		TrackID: mainTrack.ID,
		Notes:   "Second workout",
		Sections: []domain.WorkoutSection{
			{
				Title: "Разминка",
				Protocol: domain.Protocol{
					Type:        domain.ProtocolTypeCustom,
					Title:       "3 раунда",
					Description: "",
				},
				Exercises: []domain.WorkoutExercise{
					{
						ExerciseSlug: domain.ExerciseSlug("table"),
						Description:  "10 столов",
					},
					{
						ExerciseSlug: domain.ExerciseSlug("forward-bend"),
						Description:  "10 наклонов вперед",
					},
					{
						ExerciseSlug: domain.ExerciseSlug("plank-with-jumping-jack"),
						Description:  "20 качающихся планок",
					},
				},
			},
			{
				Title: "Комплекс",
				Protocol: domain.Protocol{
					Type:        domain.ProtocolTypeCustom,
					Title:       "5 раундов",
					Description: "",
				},
				Exercises: []domain.WorkoutExercise{
					{
						ExerciseSlug: domain.ExerciseSlug("deadlift-on-one-leg"),
						Description:  "24 становых на одной",
					},
					{
						ExerciseSlug: domain.ExerciseSlug("squats"),
						Description:  "18 приседаний",
					},
					{
						ExerciseSlug: domain.ExerciseSlug("push-up-with-hands-over-head"),
						Description:  "12 отжиманий",
					},
				},
			},
		},
	})
	if err != nil {
		logger.Fatal().Err(err).Msg("Failed to create workout")
	}

	logger.Info().Msg("Database successfully populated!")
}
